extensions/amp-subscriptions-google/0.1/amp-subscriptions-google.js

```diff
@@ -571,70 +617,205 @@ export class GoogleSubscriptionsPlatform {
+   * @return {Entitlement}
+   */
+  createAmpEntitlement(swgEntitlements) {
+    // Get and store the isReadyToPay signal which is independent of
+    // any entitlments existing.
+    if (swgEntitlements.isReadyToPay) {
+      this.isReadyToPay_ = true;
+    }
+
+    let swgEntitlement = swgEntitlements.getEntitlementForThis();
+    let granted = false;
+    if (swgEntitlement && swgEntitlement.source) {
+      granted = true;
+    } else if (
+      swgEntitlements.entitlements.length &&
+      swgEntitlements.entitlements[0].products.length
+    ) {
+      // We didn't find a grant so see if there is a non granting
+      // and return that. Note if we start returning multiple non
+      // granting we'll need to refactor to handle returning an
+      // array of Entitlement objects.
+      // #TODO(jpettitt) - refactor to handle multi entitlement case
+      swgEntitlement = swgEntitlements.entitlements[0];
+    } else {
+      return null;
+    }
+    swgEntitlements.ack();
+
+    // Determine grant reason.
+    let grantReason;
+    if (granted) {
+      if (swgEntitlement.source === 'google:metering') {
+        grantReason = GrantReason.METERING;
+      } else {
+        grantReason = GrantReason.SUBSCRIBER;
+      }
+    } else {
+      grantReason = null;
+    }
+
+    return new Entitlement({
+      source: swgEntitlement.source,
+      raw: swgEntitlements.raw,
+      service: PLATFORM_KEY,
+      granted,
+      // if it's granted it must be a subscriber
+      grantReason,
+      dataObject: swgEntitlement.json(),
+      decryptedDocumentKey: swgEntitlements.decryptedDocumentKey,
+    });
+  }
+
   /** @override */
   getPlatformKey() {
-    return PLATFORM_ID;
+    return PLATFORM_KEY;
   }
 
   /** @override */
-  activate(entitlement, grantEntitlement) {
+  activate(entitlement, grantEntitlement, continueAuthorizationFlow) {
     const best = grantEntitlement || entitlement;
-    // Offers or abbreviated offers may need to be shown depending on
-    // whether the access has been granted and whether user is a subscriber.
-    if (!best.granted) {
-      this.runtime_.showOffers({list: 'amp'});
-    } else if (!best.isSubscriber()) {
-      this.runtime_.showAbbrvOffer({list: 'amp'});
-    }
+
+    const googleMeteringStrategyPromise = this.getGoogleMeteringStrategy_();
+    const meteringStatePromise = this.serviceAdapter_.loadMeteringState();
+    const promises = Promise.all([
+      googleMeteringStrategyPromise,
+      meteringStatePromise,
+    ]);
+
+    promises.then((results) => {
+      const meteringStrategy = results[0];
+      const meteringState = results[1];
+
+      if (meteringStrategy === GoogleMeteringStrategy.EXTENDED_ACCESS) {
+        // Show the Regwall, so the user can get
+        // a metering state that leads to a
+        // granting entitlement.
+        // After the Regwall flow completes, then continue authorization flow.
+        if (!best.granted && !meteringState) {
```

Comment:
If an anonymous user came to a Publisher with SwG & Extended Access, wouldn't this currently show both the Regwall and showOffers?

Reply:
@ChrisAntaki - are you able to confirm on this one?

Reply:
Just one should happen at a time. The `return` statement should make sure of that

```js
if (!best.granted && !meteringState) {
  this.showMeteringRegwall_().then(continueAuthorizationFlow);
  return;
}
```